[blocked] Link to proposed LLM policy (#2835)
rust-lang/rustc-dev-guide: src/llm-guidance/writing.md

## Rules

### Before you write code

Before anything else, find a reviewer who volunteers to review your PR.
If you do not know where to find a reviewer, ask in [#llm-mentoring] on Zulip.
Your first message should say:

- your relevant experience, so we can find an issue that's suitable for you
- which problem (or kind of problem) you want to work on
- (optional) ideas you have so far for a solution
- (optional) how you expect to test your solution

Mentors are here to help.
Talking to them early helps you avoid wasted work.

### While working

Write your own doc-comments, `// SAFETY` comments, diagnostic wording, and soundness-critical code.
As before, you can use an LLM to review your work, but not to write it from scratch.
If you don't know what counts as soundness-critical, discuss it with your reviewer.

### When opening a PR

Disclose your use of LLMs, following the disclosure guidelines below.
Write the disclosure yourself.
You may use an LLM to privately review a disclosure you have written, but not to draft or rewrite it.

**Write your own PR description and comments**.
We want to hear from you, not from your agent.
LLM-created PR descriptions are banned.

> LLM-created Github comments are banned.

Reviewer: Up in the "reviewing llm code", it is said 
```
## Reviewing LLM-created code

First, add the new `ai-assisted` label to the PR.
```
can we include this prose here additionally/instead? Adding an ai-assisted label yourself feels like part of disclosure to me.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustc-dev-guide/2835/f5333a6ab8e34b445794cf7d572806765482598b..937ffd981c62885addcd293e0ba9daf90bd8f707)*

Author: this is mildly annoying for people who aren’t in the org because they have to go through triagebot to change labels. i think  disclosure in the PR description is enough, maybe we can have triagebot do something fancy with the checkboxes idk.

- What is the original bug? When does it happen? How severe is it? What causes it?
- Why is this the right fix? Are there other fixes possible? What are their advantages or disadvantages?
- Are there any edge cases? Does your code handle them?
- Does the code have existing [invariants](https://brooker.co.za/blog/2023/07/28/ds-testing.html)?
  Did you preserve those invariants?
- What behavior is *unchanged*? What test establishes that?
- Why does your test trigger the bug?
- What are you still not certain about?

It's ok to be uncertain and to ask for help.
We would much rather help you because you're not sure than have you guess wrong and then have to reverse-engineer where you went wrong.

[`ast-grep`]: https://astgrep.com/

### Disclosure guidelines

Disclose the *extent* and *purpose* of your LLM use.
We don't care which model you used, but we do care whether you used the LLM to implement the idea or to come up with it.

**Good** examples:

> LLM disclosure: I wrote the three commits by hand after viewing profiling data. I used an LLM to review the commits before submitting. The LLM identified that `ImplString::is_negative` was no longer used, so I removed that field by hand.

> > Created with the help of Claude Code, which:

Reviewer 2: ```suggestion
> The LLM helped in this way:
```

I am somehow against suggesting people to mention which LLM they used. It's free "advertisement", it's not relevant for us.

Author: sure, done

# Working with LLMs

This is a list of guidelines for working with LLMs, as well as a summary of the moderation policy.
This is not the policy itself; see [Forge][LLM policy].
If the two conflict, Forge is canonical.

[LLM policy]: https://forge.rust-lang.org/policies/llm-usage.html

## Writing LLM-created code

LLMs are a valuable tool, but one that is easy to misuse.
The main risks are **overwhelming volume** and **lack of understanding**.
When writing your PR, keep in mind that there is a person on the other end who needs to review and understand your change.
Help us out by making your change small, targeted, and easy to review.

Keep in mind this quote:

> Programs must be written for people to read, and only incidentally for machines to execute.
> — Harold Abelson and Gerald Jay Sussman, [*Structure and Interpretation of Computer Programs*][sicp]

[sicp]: https://web.mit.edu/6.001/6.037/sicp.pdf

### Rules

> Before anything else, find a reviewer who volunteers to review your PR.

Reviewer 3: (disclaimer: english is my second language, i checked the wiktionary and it seems to agree with me, but i wouldn't be surprised if i was wrong somehow)

_volunteer_ signals the review work is done without being paid, which is not always true. I would go with:

```suggestion
Before anything else, find a reviewer who is willing to review your PR.
```

Style-wise, the sentence is a bit circular, but so is the original version. I'm not sure how to address that.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustc-dev-guide/2835/f5333a6ab8e34b445794cf7d572806765482598b..32dca76297ce06c9507f9fc061fa98ae4b37f00d)*

Author: i think this is correct. i mean "volunteer" in the sense of [the verb](https://dictionary.cambridge.org/dictionary/english/volunteer), not the noun:

> a person who does something, especially helping other people, willingly and without being forced or paid to do it

Reviewer 3: it indeed is, i stand corrected!

If you do not know where to find a reviewer, ask in [#llm-mentoring] on Zulip.
Your first message should say:

- your relevant experience, so we can find an issue that's suitable for you
- which problem (or kind of problem) you want to work on
- (optional) ideas you have so far for a solution
- (optional) how you expect to test your solution

Disclose your use of LLMs, following the disclosure guidelines below.
Write the disclosure yourself.
You may use an LLM to privately review a disclosure you have written, but not to draft or rewrite it.

Write your own doc-comments, `// SAFETY` comments, diagnostic wording, and soundness-critical code.
As before, you can use an LLM to review your work, but not to write it from scratch.
If you don't know what counts as soundness-critical, discuss it with your reviewer.

**Write your own PR description and comments**.
LLM-created PR descriptions are banned.
LLM-created Github comments are banned.

> We want to hear from you, not from your agent.

Reviewer 3: I feel like putting the last sentence at the beginning of the paragraph helps a lot with people who (like me, unfortunately) read the beginning of paragraphs and tend to skip them entirely if they feel like they understand what it is about. The first sentence gives the general rule, the second sentence explains why, and the two last sentences explicitly list what's banned, this sounds much more natural to me :sparkles:

```suggestion
**Write your own PR description and comments**.
We want to hear from you, not from your agent.
LLM-created PR descriptions are banned.
LLM-created Github comments are banned.
```

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rustc-dev-guide/2835/f5333a6ab8e34b445794cf7d572806765482598b..32dca76297ce06c9507f9fc061fa98ae4b37f00d)*
